### ISSUE TITLE
[APEX] Test asserts in other methods not detected

### ISSUE THREAD

**Issue by yippie** [pmd/pmd#1089]
Affects PMD Version: 6.3

Rule: ApexUnitTestClassShouldHaveAsserts

Description: if asserts on not directly inside the test method, the method will be flagged as having no asserts even if the asserts exist in a called method

Code Sample demonstrating the issue:

@isTest
private class MyTestClass {
	
    @isTest
    static void itShouldDoTheThing() { //This gets flagged as having no assert
        //some action
        checkMyWork();
    }

    static void checkMyWork() {
        System.assert(true);
    }
}

Running PMD through: CLI

**Comment by rsoesemann** [comment]
PMD and other static analysis tools are not capable to follow the execution flow of methods as they take the source code only. This will never work and you should take the issues from PMD only as a first indication of problems. @SupressWarnings whenever you get such false positives.

**Comment by mkhokhlov** [comment]
>  This will never work

I may be missing something, but looking at the [implementation](https://github.com/pmd/pmd/blob/1a6d60aead4998a8ffca3076d8180527ce78f570/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestClassShouldHaveAssertsRule.java#L54), is it really impossible to traverse other methods collected in `methodCalls`?

That said, I can totally see doing so increasing analysis time and having to come to some compromises (e.g. supporting asserts only in the same class and only one hop from a method annotated `@isTest` .

**Comment by rsoesemann** [comment]
@SCWells72 yes this is still true. Not because of Jorje but because of the way PMD works. PMD is a single file processor. There are multiple people trying to solve that for PMD overall (Dataflow PoC will be deprecated with 7.0 see https://github.com/pmd/pmd/issues/2124 or only Salesforce. The most recent thread were we discuss this can be found here https://github.com/pmd/pmd/issues/2717.

If you want to help to overcome this please lets meet and/or continue the discussion here.

**Comment by rsoesemann** [comment]
Far out @SCWells72. Great design that improves the situation a lot. @mkhokhlov this also should solve your problem partly, right? Any improvement ideas?
